File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSHLSAction.kt
package live.hms.hmssdk_flutter

import android.util.Log
import live.hms.video.sdk.HMSSDK
import live.hms.video.sdk.models.HMSHLSConfig
import live.hms.video.sdk.models.HMSHLSMeetingURLVariant
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.video.sdk.models.HMSHlsRecordingConfig

class HMSHLSAction {

    companion object{

        fun hlsActions(call: MethodCall, result: Result,hmssdk:HMSSDK){
            when (call.method) {
                "hls_start_streaming"->{
                    hlsStreaming(call, result,hmssdk)
                }
                "hls_stop_streaming"->{
                    stopHLSStreaming(call,result,hmssdk)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }


        private fun hlsStreaming(call: MethodCall, result: Result,hmssdk:HMSSDK) {
            val meetingUrlVariantsList = call.argument<List<Map<String,String>>?>("meeting_url_variants")
            val recordingConfig = call.argument<Map<String,Boolean>?>("recording_config")
            var meetingUrlVariant : ArrayList<HMSHLSMeetingURLVariant>? = null
            var hmsHLSRecordingConfig: HMSHlsRecordingConfig? = null
            var hlsConfig: HMSHLSConfig? = null
            if(meetingUrlVariantsList!=null) {
                meetingUrlVariant = ArrayList()
                meetingUrlVariantsList?.forEach {
                    meetingUrlVariant.add(
                        HMSHLSMeetingURLVariant(
                            meetingUrl = it["meeting_url"]!!,
                            metadata = it["meta_data"]!!
                        )
                    )
                }
            }
            if(recordingConfig!=null){
                hmsHLSRecordingConfig = HMSHlsRecordingConfig(
                    singleFilePerLayer = recordingConfig?.get("single_file_per_layer")!!,
                    videoOnDemand = recordingConfig?.get("video_on_demand")!!
                )
            }
            if(meetingUrlVariant!=null || hmsHLSRecordingConfig!=null) {
                hlsConfig = HMSHLSConfig(meetingUrlVariant, hmsHLSRecordingConfig)
            }
            hmssdk.startHLSStreaming(config = hlsConfig, hmsActionResultListener = HMSCommonAction.getActionListener(result))
        }


        private fun stopHLSStreaming(call: MethodCall,result: Result,hmssdk:HMSSDK) {
            val meetingUrlVariantsList = call.argument<List<Map<String,String>>>("meeting_url_variants")

            val meetingUrlVariant1 : ArrayList<HMSHLSMeetingURLVariant> = ArrayList()

            meetingUrlVariantsList?.forEach {
                meetingUrlVariant1.add(
                        HMSHLSMeetingURLVariant(
                                meetingUrl = it["meeting_url"]!!,
                                metadata = it["meta_data"]!!
                        )
                )
            }

            var hlsConfig : HMSHLSConfig? = null
            if(meetingUrlVariant1.isNotEmpty())
                hlsConfig = HMSHLSConfig(meetingUrlVariant1)

            hmssdk.stopHLSStreaming(config = hlsConfig, hmsActionResultListener = HMSCommonAction.getActionListener(result))
        }

    }
}

File: ios/Classes/Models/HMSHLSAction.swift
//
//  HMSHLSAction.swift
//  hmssdk_flutter
//
//  Created by govind on 03/02/22.
//

import Foundation
import HMSSDK

class HMSHLSAction {
    static func hlsActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK:HMSSDK?) {
        switch call.method {
        case "hls_start_streaming":
            startHlsStreaming(call, result, hmsSDK)
            
        case "hls_stop_streaming":
            stopHlsStreaming(call, result, hmsSDK)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    static private func startHlsStreaming(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK:HMSSDK?) {
        let arguments = call.arguments as! [AnyHashable: Any]
        let meetingUrlVariantsList = arguments["meeting_url_variants"] as? [[String: String]]? ?? nil
        let recordingConfig = arguments["recording_config"] as? [String:Bool]? ?? nil
       
        var meetingUrlVariant: [HMSHLSMeetingURLVariant]? = nil
        var hmsHLSRecordingConfig: HMSHLSRecordingConfig? = nil
        
        if meetingUrlVariantsList != nil
        {
            meetingUrlVariant = [HMSHLSMeetingURLVariant]()
            meetingUrlVariantsList!.forEach { meetingUrlVariant!.append(HMSHLSMeetingURLVariant(meetingURL: URL(string: $0["meeting_url"]!)!, metadata: $0["meta_data"] ?? "")) }
        }
        
        if recordingConfig != nil
        {
            hmsHLSRecordingConfig = HMSHLSRecordingConfig(singleFilePerLayer: recordingConfig!["single_file_per_layer"]!, enableVOD: recordingConfig!["video_on_demand"]!)
        }
        
        var hlsConfig:HMSHLSConfig? = nil
        if(meetingUrlVariant != nil || hmsHLSRecordingConfig != nil){
            hlsConfig = HMSHLSConfig(variants: meetingUrlVariant,recording: hmsHLSRecordingConfig)
        }
        hmsSDK?.startHLSStreaming(config: hlsConfig) { _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }
    
    static private func stopHlsStreaming(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        let arguments = call.arguments as? [AnyHashable: Any]
        let config = getHLSConfig(from: arguments)
        hmsSDK?.stopHLSStreaming(config: config) { _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }
    
    static private func getHLSConfig(from arguments: [AnyHashable: Any]?) -> HMSHLSConfig? {
        guard let meetingUrlVariantsList = arguments?["meeting_url_variants"] as? [[String: String]] else {
            return nil
        }
        var meetingUrlVariant = [HMSHLSMeetingURLVariant]()
        meetingUrlVariantsList.forEach { meetingUrlVariant.append(HMSHLSMeetingURLVariant(meetingURL: URL(string: $0["meeting_url"]!)!, metadata: $0["meta_data"] ?? "")) }
        
        return HMSHLSConfig(variants: meetingUrlVariant)
    }
}
